File: app/src/main/java/com/nvvi9/giphyme/ui/adapters/GifItemAdapter.java
package com.nvvi9.giphyme.ui.adapters;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.paging.PagingDataAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.card.MaterialCardView;
import com.nvvi9.giphyme.R;
import com.nvvi9.giphyme.databinding.ItemGifBinding;
import com.nvvi9.giphyme.ui.listeners.GifItemListener;
import com.nvvi9.giphyme.vo.GifItem;

public class GifItemAdapter extends PagingDataAdapter<GifItem, GifItemAdapter.GifItemViewHolder> {

    private final GifItemListener listener;

    public GifItemAdapter(GifItemListener listener) {
        super(new GifItemComparator());
        this.listener = listener;
    }

    @NonNull
    @Override
    public GifItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new GifItemViewHolder(ItemGifBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false), listener);
    }

    @Override
    public void onBindViewHolder(@NonNull GifItemViewHolder holder, int position) {
        GifItem item = getItem(position);
        if (item != null) {
            holder.bind(item);
        }
    }

    static class GifItemViewHolder extends RecyclerView.ViewHolder {

        private final float addedCornerSize = itemView.getResources().getDimension(R.dimen.small_component_corner_radius);
        private final ItemGifBinding binding;

        public GifItemViewHolder(@NonNull ItemGifBinding binding, GifItemListener listener) {
            super(binding.getRoot());
            this.binding = binding;
            this.binding.setListener(listener);
        }

        public void bind(GifItem gifItem) {
            binding.setGifItem(gifItem);
            binding.getRoot().setActivated(gifItem.getInFavourites());
            updateCardViewBottomLeftCornerSize(gifItem.getInFavourites() ? 1f : 0f);
            binding.executePendingBindings();
        }

        private void updateCardViewBottomLeftCornerSize(float interpolation) {
            MaterialCardView cardView = binding.cardView;
            cardView.setShapeAppearanceModel(
                    cardView.getShapeAppearanceModel().toBuilder()
                            .setBottomLeftCornerSize(interpolation * addedCornerSize)
                            .build()
            );
        }
    }

    private static class GifItemComparator extends DiffUtil.ItemCallback<GifItem> {

        @Override
        public boolean areItemsTheSame(@NonNull GifItem oldItem, @NonNull GifItem newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull GifItem oldItem, @NonNull GifItem newItem) {
            return oldItem.equals(newItem);
        }
    }
}

File: app/src/main/java/com/nvvi9/giphyme/ui/adapters/GifItemAdapter.kt
package com.nvvi9.giphyme.ui.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.nvvi9.giphyme.R
import com.nvvi9.giphyme.databinding.ItemGifBinding
import com.nvvi9.giphyme.ui.adapters.GifItemAdapter.GifItemViewHolder
import com.nvvi9.giphyme.ui.listeners.GifItemListener
import com.nvvi9.giphyme.vo.GifItem

class GifItemAdapter(private val listener: GifItemListener) :
    PagingDataAdapter<GifItem, GifItemViewHolder>(GifItemComparator) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GifItemViewHolder =
        GifItemViewHolder(
            ItemGifBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            ), listener
        )

    override fun onBindViewHolder(holder: GifItemViewHolder, position: Int) {
        getItem(position)?.let {
            holder.bind(it)
        }
    }

    class GifItemViewHolder(
        private val binding: ItemGifBinding,
        listener: GifItemListener
    ) : RecyclerView.ViewHolder(
        binding.root
    ) {
        private val addedCornerSize =
            itemView.resources.getDimension(R.dimen.small_component_corner_radius)

        init {
            binding.listener = listener
        }

        fun bind(gifItem: GifItem) {
            binding.gifItem = gifItem
            binding.root.isActivated = gifItem.inFavourites
            updateCardViewBottomLeftCornerSize(if (gifItem.inFavourites) 1f else 0f)
            binding.executePendingBindings()
        }

        private fun updateCardViewBottomLeftCornerSize(interpolation: Float) {
            val cardView = binding.cardView
            cardView.shapeAppearanceModel = cardView.shapeAppearanceModel.toBuilder()
                .setBottomLeftCornerSize(interpolation * addedCornerSize)
                .build()
        }
    }

    object GifItemComparator : DiffUtil.ItemCallback<GifItem>() {
        override fun areItemsTheSame(oldItem: GifItem, newItem: GifItem): Boolean =
            oldItem.id == newItem.id

        override fun areContentsTheSame(oldItem: GifItem, newItem: GifItem): Boolean =
            oldItem == newItem
    }
}